### ISSUE TITLE
decode_poyline() unrecoverable crash for invalid input

### ISSUE THREAD

**Issue by mattiZed** [urschrei/pypolyline#60]
Hi!

First of all, thanks for this awesome package!

Steps to reproduce:

```
Python 3.8.18 (default, Sep 15 2023, 08:42:09) 
[Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pypolyline
>>> pypolyline.__version__
'0.3.6'
>>> from pypolyline.cutil import decode_polyline
>>> decode_polyline(b"invalid_polyline_that_should_be_handled_gracefully", 6)
thread '<unnamed>' panicked at 'index out of bounds: the len is 50 but the index is 50', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/polyline-0.10.0/src/lib.rs:143:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
zsh: abort      python3
```

We are using `pypolyline`'s cython module and have built error handling around it by checking if the first pair of coordinates is `[nan, nan]` or not, as stated [here](https://github.com/urschrei/pypolyline?tab=readme-ov-file#cython-module-). This has worked **fine** in the past.

Currently, with release 0.3.6 (I can not exactly pinpoint since which version one might experience this problem, but I'm certain that it has not happened with "some" earlier version, probably < 0.3), we run into an unrecoverable error: the rust binary crashes with an out-of-bounds error and this in turn causes the python interpreter to terminate as well.

I have done some digging (the Rust error is very helpful), and the issue is with this line:

https://github.com/georust/polyline/blob/v0.10.0/src/lib.rs#L143

While decoding the coordinate deltas, the only termination condition for that loop is the value of a decoded byte, so if the polyline is invalid, an out-of-bounds access into the `chars` buffer may well happen.

I have done some more digging and found that this part of the rust code actually has been subject to a refactor, and before it looked like this:

https://github.com/georust/polyline/blob/v0.7.1/src/lib.rs#L121

In other words: the `georust` crate seems to be lacking proper error handling/out-of-bounds protection in this location and the bug was introduced into `pypolyline` package with bumping to `georust` v0.10.0.

I have seen that you contribute to georust crate as well.

What do you propose would be the best way going forward?

IMHO this needs fixing in `georust` crate and I could open an issue there as well, linking to this one.
But as you are the maintainer of the `pypolyline` package, I posted here first.

Any help much appreciated!

Best,
Stefan

**Comment by urschrei** [comment]
> IMHO this needs fixing in georust crate and I could open an issue there as well, linking to this one.

Yep, this is an error in `polyline`, please open the underlying issue there. Closing in favour of that.

**Comment by mattiZed** [comment]
Hi, thanks for the reply.

For reference, here is the issue i filed in [georust/poyline](https://github.com/georust/polyline/issues/36).

**Comment by urschrei** [comment]
@mattiZed Can you try the appropriate wheel for your Python version and platform from here: https://github.com/urschrei/pypolyline/releases/tag/v0.3.11? I've implemented a panic catch until we can fix the issue upstream.
